### ISSUE TITLE
Problems of running time and microbatch ?

### ISSUE THREAD

**Issue by Zecheng-Tang** [Shark-NLP/DiffuSeq#5]
Thanks for your great work. When we reproduce your results, we come up two questions:
1. The training time of  QQP task is much longer than 2 days (48h), which is inconsistent with the statement in this repo:
> It will take around 2 days to train a DiffuSeq model on 2 NVIDIA A100 GPUs for QG and QQP

We follow the suggestion implementation: 2 Nvidia A100 GPUs and run the following command: 
```python
python -m torch.distributed.launch --nproc_per_node=4 --master_port=12233 --use_env run_train.py --diff_steps 2000 --lr 0.0001 --learning_steps 50000 --microbatch 64 --save_interval 10000 --seed 102 --noise_schedule sqrt --hidden_dim 128 --bsz 2048 --dataset qqp --data_dir {datasets/QQP} --vocab bert --seq_len 128 --schedule_sampler lossaware --notes qqp
```
Here is our wandb output:
![f2df1ad229d2d1fe083be2b76b01005](https://user-images.githubusercontent.com/59752775/203501861-aadee87a-328f-443c-a56f-4727de1a5437.jpg)
as well as the log out:
![77ad9e776687786c2bacaec5c0acc0d](https://user-images.githubusercontent.com/59752775/203501970-16c56765-0b79-4a6b-8f31-75d36c236d71.jpg)
Note that we have already run this code with around 15 hours, but just reach 4960 step, which is far from 50000 steps as you suggest.

2. We take a deeper look into your code and find that during the training loops you split one batch into several micro bateches.
We wonder why you conduct such operation? Is this trick necessary?

Also, we wonder how can the total training process can be done into 2 days with your suggestion command and implementation ?

**Comment by summmeer** [comment]
Hi, 
according to your log output, it seems that your code just went through one batch. The batch and micro batch setting is not a necessary trick, just make sure the same setting with our development code. I also tried to set the `batchsize=350` and `microbatch=350`(with 2GPUs), the results were similar. So it is not this setting that causes your problem. I guess you didn't allocate enough GPUs? In the script, `--nproc_per_node=4`, it should be 4 GPUs.

Maybe the description misleads you. Let me correct the time consuming: 
- It will take around 2 days to train a DiffuSeq model on 4 NVIDIA A100 GPUs for QG and QQP

**Comment by Zecheng-Tang** [comment]
> Hi, according to your log output, it seems that your code just went through one batch. The batch and micro batch setting is not a necessary trick, just make sure the same setting with our development code. I also tried to set the `batchsize=350` and `microbatch=350`(with 2GPUs), the results were similar. So it is not this setting that causes your problem. I guess you didn't allocate enough GPUs? In the script, `--nproc_per_node=4`, it should be 4 GPUs.
> 
> Maybe the description misleads you. Let me correct the time consuming:
> 
> * It will take around 2 days to train a DiffuSeq model on 4 NVIDIA A100 GPUs for QG and QQP
> 
> BTW, please tell me if you fix it.

Thanks for your suggestion, we correct our code immediately and run successfully.
We hava another datailed question: have you utilized fp16 during the training ?
And do you have any plan for releasing the checkpoint of your model ?
Thanks again for your suggestion.

**Comment by TheNamelesswang** [comment]
> Hi, according to your log output, it seems everything is right. The batch and micro batch setting is not a necessary trick, just make sure the same setting with our development code. I also tried to set the `batchsize=350` and `microbatch=350`(with 2GPUs), the results were similar. So it is not this setting that causes your problem.
> 
> Maybe the description misleads you. Let me correct the time consuming:
> 
> * It will take around 2 more days to train a DiffuSeq model on 4 NVIDIA A100 GPUs for QG and QQP
> 
> BTW, please tell me if you have any updates.

Thanks for your great work. However, when the batch size is set to 350, does the value of learning_steps need to be changed ? According to the training command in README, one step should contain 2048 sentences while it contains 350 sentences currently.

**Comment by summmeer** [comment]
> have you utilized fp16 during the training ? And do you have any plan for releasing the checkpoint of your model ? 

@Zecheng-Tang We didn't use fp16 yet. I think releasing the checkpoint of the task-specific model does not mean much, but it is possible to prepare a download link if you need.

**Comment by summmeer** [comment]
> when the batch size is set to 350, does the value of learning_steps need to be changed ? According to the training command in README, one step should contain 2048 sentences while it contains 350 sentences currently.

@TheNamelesswang For QQP, according to my own experiment, it seems that the same `learning_steps` can produce close results, but I didn't tune the hyperparameters carefully, you are free to try.

**Comment by summmeer** [comment]
> We only have 40GB Nvidia A100 GPUs

@Zecheng-Tang When setting `batchsize=2046` and `microbatch=64`, I think 40GB is enough. If not, you can also try smaller batchsize. According to my experiments, the results are robust to different batchsize and microbatch.

**Comment by Zecheng-Tang** [comment]
> > We only have 40GB Nvidia A100 GPUs
> 
> @Zecheng-Tang When setting `batchsize=2046` and `microbatch=64`, I think 40GB is enough. If not, you can also try smaller batchsize. According to my experiments, the results are robust to different batchsize and microbatch.

Yes, I agree with you. When we set micro batch as 64, 40GB is actually enough. However, as I mentioned in the begining, the training time is too long, which cannot match the 2 days time mentioned in the repo, i.e., the training time is much longer than 7 days with 4 A100 GPUs. Specifically, we want to have a quick look of the inference results first and I hope to obtain the downlink (to my gmail or released in the repo).

**Comment by summmeer** [comment]
@Zecheng-Tang Theoretically the computing speed of 40GB A100 and 80GB A100 should be similar. Actually, the sampling will also cost several hours for each random seed. I prepare the ckpt and sampling results of 10 seeds in this [link](https://drive.google.com/drive/folders/1vnhJIUqPQva_x_sH2h5a0moCc1NYmEpr?usp=sharing).

**Comment by Zecheng-Tang** [comment]
@summmeer Thanks any way! We will try the model you provide immediately !
